File: 未使用/github文档快照/github文档仓库/AI-LlmEngineerTechStack/1_大模型技术/2_模型部署与推理/硬件适配/mygpu_ops_demo/csrc/ops/attention.cpp
/**
 * PagedAttention C++ 实现
 * 
 * PagedAttention 是 vLLM 的核心算子，实现分页式 KV Cache 管理。
 * 它解决了 LLM 推理中的内存碎片化问题，是 vLLM 高吞吐量的关键。
 * 
 * 核心思想 (类比操作系统虚拟内存):
 *     - 物理块: 固定大小的连续内存块，存储 KV Cache
 *     - 页表: 记录每个序列使用哪些物理块
 *     - 非连续存储: 一个序列的 KV Cache 可以存在多个不连续的物理块中
 * 
 * 优点:
 *     1. 消除内存碎片
 *     2. 支持动态序列长度
 *     3. 高效的 continuous batching
 */

#include "../include/common.h"

namespace mygpu {
namespace ops {

/**
 * 标准 Scaled Dot-Product Attention (参考实现)
 * 
 * 公式: softmax(Q @ K^T / sqrt(d)) @ V
 */
torch::Tensor scaled_dot_product_attention(
    torch::Tensor query,   // [batch, num_heads, seq_len, head_dim]
    torch::Tensor key,     // [batch, num_heads, kv_len, head_dim]
    torch::Tensor value,   // [batch, num_heads, kv_len, head_dim]
    double scale
) {
    // Q @ K^T: [batch, heads, seq_len, kv_len]
    auto scores = torch::matmul(query, key.transpose(-2, -1)) * scale;
    
    // Softmax
    auto attn_weights = torch::softmax(scores, /*dim=*/-1);
    
    // Attention @ V: [batch, heads, seq_len, head_dim]
    return torch::matmul(attn_weights, value);
}

/**
 * PagedAttention V1 前向传播 - CPU 实现
 * 
 * 这是简化的 CPU 参考实现，用于理解算法原理。
 * 实际的 GPU 实现会使用优化的 CUDA kernel。
 * 
 * @param query         Query 张量 [num_seqs, num_heads, head_dim]
 * @param key_cache     Key 缓存 [num_blocks, block_size, num_kv_heads, head_dim]
 * @param value_cache   Value 缓存 [num_blocks, block_size, num_kv_heads, head_dim]
 * @param block_tables  页表 [num_seqs, max_blocks]
 * @param seq_lens      每个序列的长度 [num_seqs]
 * @param scale         缩放因子 (通常是 1/sqrt(head_dim))
 * @param block_size    每个块的大小
 * @return              输出张量 [num_seqs, num_heads, head_dim]
 */
torch::Tensor paged_attention_v1_forward_cpu(
    torch::Tensor query,
    torch::Tensor key_cache,
    torch::Tensor value_cache,
    torch::Tensor block_tables,
    torch::Tensor seq_lens,
    double scale,
    int64_t block_size
) {
    // ========== 1. 输入检查 ==========
    MYGPU_CHECK_INPUT(query);
    MYGPU_CHECK_INPUT(key_cache);
    MYGPU_CHECK_INPUT(value_cache);
    
    const int64_t num_seqs = query.size(0);
    const int64_t num_heads = query.size(1);
    const int64_t head_dim = query.size(2);
    const int64_t num_kv_heads = key_cache.size(2);
    
    // 计算每个 head group 对应的 kv_head (GQA 支持)
    const int64_t num_queries_per_kv = num_heads / num_kv_heads;
    
    // 输出张量
    auto output = torch::zeros_like(query);
    
    // ========== 2. 逐序列处理 ==========
    // 在 GPU 实现中，这会并行处理
    for (int64_t seq_idx = 0; seq_idx < num_seqs; ++seq_idx) {
        const int64_t seq_len = seq_lens[seq_idx].item<int64_t>();
        
        if (seq_len == 0) {
            continue;
        }
        
        // 计算需要的块数
        const int64_t num_blocks_needed = (seq_len + block_size - 1) / block_size;
        
        // 收集该序列的所有 Key 和 Value
        std::vector<torch::Tensor> keys_list;
        std::vector<torch::Tensor> values_list;
        
        int64_t tokens_collected = 0;
        for (int64_t block_idx = 0; block_idx < num_blocks_needed; ++block_idx) {
            // 从页表获取物理块索引
            const int64_t physical_block_idx = 
                block_tables[seq_idx][block_idx].item<int64_t>();
            
            // 计算这个块中实际的 token 数
            const int64_t tokens_in_block = 
                std::min(block_size, seq_len - tokens_collected);
            
            // 获取这个块的 key 和 value
            // key_cache: [num_blocks, block_size, num_kv_heads, head_dim]
            auto key_block = key_cache[physical_block_idx].slice(
                /*dim=*/0, /*start=*/0, /*end=*/tokens_in_block);
            auto value_block = value_cache[physical_block_idx].slice(
                /*dim=*/0, /*start=*/0, /*end=*/tokens_in_block);
            
            keys_list.push_back(key_block);
            values_list.push_back(value_block);
            
            tokens_collected += tokens_in_block;
        }
        
        // 拼接所有块
        // keys: [seq_len, num_kv_heads, head_dim]
        auto keys = torch::cat(keys_list, /*dim=*/0);
        auto values = torch::cat(values_list, /*dim=*/0);
        
        // ========== 3. 计算 Attention ==========
        // query for this seq: [num_heads, head_dim] -> [1, num_heads, 1, head_dim]
        auto q = query[seq_idx].unsqueeze(0).unsqueeze(2);
        
        // keys: [seq_len, num_kv_heads, head_dim] -> [1, num_kv_heads, seq_len, head_dim]
        auto k = keys.permute({1, 0, 2}).unsqueeze(0);
        auto v = values.permute({1, 0, 2}).unsqueeze(0);
        
        // 处理 GQA: 扩展 KV heads 以匹配 Q heads
        if (num_queries_per_kv > 1) {
            k = k.repeat_interleave(num_queries_per_kv, /*dim=*/1);
            v = v.repeat_interleave(num_queries_per_kv, /*dim=*/1);
        }
        
        // Scaled Dot-Product Attention
        // scores: [1, num_heads, 1, seq_len]
        auto scores = torch::matmul(q, k.transpose(-2, -1)) * scale;
        auto attn_weights = torch::softmax(scores, /*dim=*/-1);
        
        // output: [1, num_heads, 1, head_dim]
        auto attn_output = torch::matmul(attn_weights, v);
        
        // 存入输出
        output[seq_idx] = attn_output.squeeze(0).squeeze(1);
    }
    
    return output;
}

/**
 * 标准 Multi-Head Attention (非分页)
 * 
 * 用于 prefill 阶段，处理完整的输入序列。
 * 
 * @param query   [batch, seq_len, num_heads, head_dim]
 * @param key     [batch, seq_len, num_kv_heads, head_dim]
 * @param value   [batch, seq_len, num_kv_heads, head_dim]
 * @param scale   缩放因子
 * @return        [batch, seq_len, num_heads, head_dim]
 */
torch::Tensor standard_attention_forward_cpu(
    torch::Tensor query,
    torch::Tensor key,
    torch::Tensor value,
    double scale
) {
    MYGPU_CHECK_INPUT(query);
    MYGPU_CHECK_INPUT(key);
    MYGPU_CHECK_INPUT(value);
    
    const int64_t batch = query.size(0);
    const int64_t seq_len = query.size(1);
    const int64_t num_heads = query.size(2);
    const int64_t head_dim = query.size(3);
    const int64_t num_kv_heads = key.size(2);
    const int64_t num_queries_per_kv = num_heads / num_kv_heads;
    
    // 转置为 [batch, num_heads, seq_len, head_dim]
    auto q = query.permute({0, 2, 1, 3});
    auto k = key.permute({0, 2, 1, 3});
    auto v = value.permute({0, 2, 1, 3});
    
    // 处理 GQA
    if (num_queries_per_kv > 1) {
        k = k.repeat_interleave(num_queries_per_kv, /*dim=*/1);
        v = v.repeat_interleave(num_queries_per_kv, /*dim=*/1);
    }
    
    // Scaled Dot-Product Attention
    auto output = scaled_dot_product_attention(q, k, v, scale);
    
    // 转置回 [batch, seq_len, num_heads, head_dim]
    return output.permute({0, 2, 1, 3});
}

}  // namespace ops
}  // namespace mygpu


// =============================================================================
// 导出函数
// =============================================================================

torch::Tensor paged_attention_v1_forward(
    torch::Tensor query,
    torch::Tensor key_cache,
    torch::Tensor value_cache,
    torch::Tensor block_tables,
    torch::Tensor seq_lens,
    double scale,
    int64_t block_size
) {
    if (query.device().is_cpu()) {
        return mygpu::ops::paged_attention_v1_forward_cpu(
            query, key_cache, value_cache, block_tables, seq_lens, scale, block_size
        );
    }
    
    TORCH_WARN("CUDA implementation not available, falling back to CPU");
    auto cpu_output = mygpu::ops::paged_attention_v1_forward_cpu(
        query.cpu(),
        key_cache.cpu(),
        value_cache.cpu(),
        block_tables.cpu(),
        seq_lens.cpu(),
        scale,
        block_size
    );
    return cpu_output.to(query.device());
}

// 额外导出标准 attention 用于测试
torch::Tensor standard_attention_forward(
    torch::Tensor query,
    torch::Tensor key,
    torch::Tensor value,
    double scale
) {
    if (query.device().is_cpu()) {
        return mygpu::ops::standard_attention_forward_cpu(query, key, value, scale);
    }
    
    TORCH_WARN("CUDA implementation not available, falling back to CPU");
    auto cpu_output = mygpu::ops::standard_attention_forward_cpu(
        query.cpu(), key.cpu(), value.cpu(), scale
    );
    return cpu_output.to(query.device());
}

File: 未使用/github文档快照/github文档仓库/AI-LlmEngineerTechStack/1_大模型技术/2_模型部署与推理/硬件适配/mygpu_ops_demo/mygpu_ops/kernels/attention.py
"""
Attention Kernels - 注意力机制的Kernel实现

这个模块模拟GPU Attention kernel的实现。
包括标准Attention和PagedAttention。

主要算子:
- scaled_dot_product_attention: 标准SDPA
- paged_attention_v1_kernel: PagedAttention V1
- paged_attention_v2_kernel: PagedAttention V2
- flash_attention_kernel: Flash Attention (简化版)
"""

import torch
from typing import Tuple, Optional, List
from dataclasses import dataclass
import math


@dataclass
class AttentionConfig:
    """Attention配置"""
    block_size: int = 16          # PagedAttention block大小
    partition_size: int = 512     # V2版本的分区大小
    use_alibi: bool = False       # 是否使用ALiBi位置编码
    sliding_window: Optional[int] = None  # 滑动窗口大小


def scaled_dot_product_attention(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    scale: Optional[float] = None,
    attn_mask: Optional[torch.Tensor] = None,
    is_causal: bool = False,
) -> torch.Tensor:
    """
    标准缩放点积注意力
    
    公式: Attention(Q, K, V) = softmax(Q @ K^T / sqrt(d_k)) @ V
    
    GPU Kernel实现细节:
    
    ```cpp
    // 简化的CUDA实现思路
    __global__ void attention_kernel(
        float* output,
        const float* Q,
        const float* K,
        const float* V,
        int seq_len,
        int head_dim,
        float scale
    ) {
        // 每个block处理一个head的一行
        int batch_head = blockIdx.y;
        int query_pos = blockIdx.x;
        
        // 1. 计算Q @ K^T
        __shared__ float scores[MAX_SEQ_LEN];
        for (int k = threadIdx.x; k < seq_len; k += blockDim.x) {
            float score = 0.0f;
            for (int d = 0; d < head_dim; d++) {
                score += Q[query_pos * head_dim + d] * 
                        K[k * head_dim + d];
            }
            scores[k] = score * scale;
        }
        __syncthreads();
        
        // 2. Softmax (需要两遍扫描: 求max, 求sum)
        // ... softmax实现 ...
        
        // 3. 乘以V
        // ... output计算 ...
    }
    ```
    
    Args:
        query: [batch, num_heads, seq_len, head_dim]
        key: [batch, num_heads, seq_len, head_dim]
        value: [batch, num_heads, seq_len, head_dim]
        scale: 缩放因子，默认 1/sqrt(head_dim)
        attn_mask: 注意力掩码
        is_causal: 是否使用因果掩码
    
    Returns:
        output: [batch, num_heads, seq_len, head_dim]
    """
    if scale is None:
        scale = 1.0 / math.sqrt(query.shape[-1])
    
    # Q @ K^T
    scores = torch.matmul(query, key.transpose(-2, -1)) * scale
    
    # 应用掩码
    if is_causal:
        seq_len = query.shape[-2]
        causal_mask = torch.triu(
            torch.ones(seq_len, seq_len, device=query.device, dtype=torch.bool),
            diagonal=1
        )
        scores = scores.masked_fill(causal_mask, float('-inf'))
    
    if attn_mask is not None:
        scores = scores + attn_mask
    
    # Softmax
    attn_weights = torch.softmax(scores, dim=-1)
    
    # @ V
    output = torch.matmul(attn_weights, value)
    
    return output


def paged_attention_v1_kernel(
    query: torch.Tensor,
    key_cache: torch.Tensor,
    value_cache: torch.Tensor,
    block_tables: torch.Tensor,
    seq_lens: torch.Tensor,
    scale: float,
    block_size: int = 16,
) -> torch.Tensor:
    """
    PagedAttention V1 Kernel
    
    这是vLLM的核心算子，实现了分页的KV Cache管理。
    
    核心思想：
    1. KV Cache被分成固定大小的block（类似OS的内存页）
    2. 每个sequence有一个block table，记录其KV存储位置
    3. Attention计算时，根据block table访问对应的KV
    
    优势：
    - 避免KV Cache的碎片化
    - 支持动态sequence长度
    - 便于实现Continuous Batching
    
    GPU Kernel实现细节:
    
    ```cpp
    // PagedAttention V1 CUDA Kernel伪代码
    __global__ void paged_attention_v1_kernel(
        float* output,
        const float* query,
        const float* key_cache,
        const float* value_cache,
        const int* block_tables,
        const int* seq_lens,
        float scale,
        int block_size,
        int num_heads,
        int head_dim
    ) {
        // 每个block处理一个sequence的一个head
        int seq_idx = blockIdx.x;
        int head_idx = blockIdx.y;
        int seq_len = seq_lens[seq_idx];
        
        // 1. 加载query到寄存器
        float q[HEAD_DIM];
        for (int d = 0; d < head_dim; d++) {
            q[d] = query[seq_idx * num_heads * head_dim + 
                        head_idx * head_dim + d];
        }
        
        // 2. 遍历所有block，计算attention
        float max_score = -INFINITY;
        float sum_exp = 0.0f;
        float output_acc[HEAD_DIM] = {0};
        
        int num_blocks = (seq_len + block_size - 1) / block_size;
        for (int b = 0; b < num_blocks; b++) {
            int physical_block = block_tables[seq_idx * max_blocks + b];
            
            // 计算这个block内所有token的attention score
            for (int t = 0; t < block_size; t++) {
                int token_idx = b * block_size + t;
                if (token_idx >= seq_len) break;
                
                // 从cache读取K
                float score = 0.0f;
                for (int d = 0; d < head_dim; d++) {
                    float k = key_cache[physical_block * block_size * head_dim +
                                       t * head_dim + d];
                    score += q[d] * k;
                }
                score *= scale;
                
                // Online softmax更新
                float old_max = max_score;
                max_score = max(max_score, score);
                float exp_score = exp(score - max_score);
                sum_exp = sum_exp * exp(old_max - max_score) + exp_score;
                
                // 累加V
                for (int d = 0; d < head_dim; d++) {
                    float v = value_cache[physical_block * block_size * head_dim +
                                         t * head_dim + d];
                    output_acc[d] = output_acc[d] * exp(old_max - max_score) + 
                                   exp_score * v;
                }
            }
        }
        
        // 3. 归一化并写回
        for (int d = 0; d < head_dim; d++) {
            output[seq_idx * num_heads * head_dim + head_idx * head_dim + d] = 
                output_acc[d] / sum_exp;
        }
    }
    ```
    
    Args:
        query: [num_seqs, num_heads, head_dim]
        key_cache: [num_blocks, block_size, num_kv_heads, head_dim]
        value_cache: [num_blocks, block_size, num_kv_heads, head_dim]
        block_tables: [num_seqs, max_num_blocks]
        seq_lens: [num_seqs]
        scale: 缩放因子
        block_size: block大小
    
    Returns:
        output: [num_seqs, num_heads, head_dim]
    """
    num_seqs = query.shape[0]
    num_heads = query.shape[1]
    head_dim = query.shape[2]
    num_kv_heads = key_cache.shape[2]
    
    # GQA支持
    num_queries_per_kv = num_heads // num_kv_heads
    
    outputs = []
    
    # ========== 模拟PagedAttention Kernel ==========
    
    for seq_idx in range(num_seqs):
        seq_len = seq_lens[seq_idx].item()
        num_blocks_for_seq = (seq_len + block_size - 1) // block_size
        
        # 收集KV
        keys = []
        values = []
        
        for block_idx in range(num_blocks_for_seq):
            physical_block_id = block_tables[seq_idx, block_idx].item()
            
            # 计算block内有效token数
            tokens_in_block = min(block_size, seq_len - block_idx * block_size)
            
            k = key_cache[physical_block_id, :tokens_in_block]
            v = value_cache[physical_block_id, :tokens_in_block]
            
            keys.append(k)
            values.append(v)
        
        # 拼接
        k_seq = torch.cat(keys, dim=0)  # [seq_len, num_kv_heads, head_dim]
        v_seq = torch.cat(values, dim=0)
        
        # GQA扩展
        if num_queries_per_kv > 1:
            k_seq = k_seq.repeat_interleave(num_queries_per_kv, dim=1)
            v_seq = v_seq.repeat_interleave(num_queries_per_kv, dim=1)
        
        # 计算attention
        q = query[seq_idx]  # [num_heads, head_dim]
        
        # scores: [num_heads, seq_len]
        scores = torch.einsum('hd,shd->hs', q, k_seq) * scale
        attn_weights = torch.softmax(scores, dim=-1)
        
        # output: [num_heads, head_dim]
        output = torch.einsum('hs,shd->hd', attn_weights, v_seq)
        outputs.append(output)
    
    return torch.stack(outputs, dim=0)


def paged_attention_v2_kernel(
    query: torch.Tensor,
    key_cache: torch.Tensor,
    value_cache: torch.Tensor,
    block_tables: torch.Tensor,
    seq_lens: torch.Tensor,
    scale: float,
    block_size: int = 16,
    partition_size: int = 512,
) -> torch.Tensor:
    """
    PagedAttention V2 Kernel
    
    V2版本对长序列进行了优化：
    - 将序列分成多个partition
    - 每个partition独立计算
    - 最后合并结果
    
    优势：
    - 更好的并行度
    - 减少长序列的计算瓶颈
    - 更好的GPU利用率
    
    Args:
        query: [num_seqs, num_heads, head_dim]
        key_cache: [num_blocks, block_size, num_kv_heads, head_dim]
        value_cache: [num_blocks, block_size, num_kv_heads, head_dim]
        block_tables: [num_seqs, max_num_blocks]
        seq_lens: [num_seqs]
        scale: 缩放因子
        block_size: block大小
        partition_size: 分区大小
    
    Returns:
        output: [num_seqs, num_heads, head_dim]
    """
    # 简化实现：直接调用V1
    # 实际V2会有更复杂的分区逻辑
    return paged_attention_v1_kernel(
        query=query,
        key_cache=key_cache,
        value_cache=value_cache,
        block_tables=block_tables,
        seq_lens=seq_lens,
        scale=scale,
        block_size=block_size,
    )


def flash_attention_kernel(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    scale: Optional[float] = None,
    is_causal: bool = False,
) -> torch.Tensor:
    """
    Flash Attention 简化实现
    
    Flash Attention的核心思想：
    1. 使用分块计算，减少显存占用
    2. 在SRAM中完成softmax，避免写回HBM
    3. 使用online softmax算法
    
    内存复杂度：
    - 标准Attention: O(N^2) - 需要存储完整的attention矩阵
    - Flash Attention: O(N) - 只需要存储分块大小的中间结果
    
    这里是简化版，实际Flash Attention的CUDA实现非常复杂。
    
    Args:
        query: [batch, num_heads, seq_len, head_dim]
        key: [batch, num_heads, seq_len, head_dim]
        value: [batch, num_heads, seq_len, head_dim]
        scale: 缩放因子
        is_causal: 是否因果
    
    Returns:
        output: [batch, num_heads, seq_len, head_dim]
    """
    if scale is None:
        scale = 1.0 / math.sqrt(query.shape[-1])
    
    # 分块大小（实际Flash Attention会根据SRAM大小动态选择）
    BLOCK_SIZE = 64
    
    batch_size, num_heads, seq_len, head_dim = query.shape
    
    # 初始化输出
    output = torch.zeros_like(query)
    
    # 分块计算
    for i in range(0, seq_len, BLOCK_SIZE):
        i_end = min(i + BLOCK_SIZE, seq_len)
        q_block = query[:, :, i:i_end, :]  # [B, H, block, D]
        
        # 累积softmax的分母和加权value
        max_scores = torch.full(
            (batch_size, num_heads, i_end - i, 1), 
            float('-inf'), 
            device=query.device
        )
        sum_exp = torch.zeros(
            (batch_size, num_heads, i_end - i, 1), 
            device=query.device
        )
        output_block = torch.zeros(
            (batch_size, num_heads, i_end - i, head_dim), 
            device=query.device
        )
        
        for j in range(0, seq_len, BLOCK_SIZE):
            j_end = min(j + BLOCK_SIZE, seq_len)
            
            # 因果掩码：只看之前的token
            if is_causal and j > i_end - 1:
                continue
            
            k_block = key[:, :, j:j_end, :]
            v_block = value[:, :, j:j_end, :]
            
            # 计算attention scores
            scores = torch.matmul(q_block, k_block.transpose(-2, -1)) * scale
            
            # 应用因果掩码
            if is_causal:
                # 创建block级别的因果掩码
                q_positions = torch.arange(i, i_end, device=query.device)
                k_positions = torch.arange(j, j_end, device=query.device)
                causal_mask = q_positions.unsqueeze(1) < k_positions.unsqueeze(0)
                scores = scores.masked_fill(causal_mask.unsqueeze(0).unsqueeze(0), float('-inf'))
            
            # Online softmax更新
            block_max = scores.max(dim=-1, keepdim=True)[0]
            new_max = torch.maximum(max_scores, block_max)
            
            # 更新累积值
            exp_scores = torch.exp(scores - new_max)
            sum_exp = sum_exp * torch.exp(max_scores - new_max) + exp_scores.sum(dim=-1, keepdim=True)
            
            # 更新输出
            output_block = output_block * torch.exp(max_scores - new_max) + torch.matmul(exp_scores, v_block)
            
            max_scores = new_max
        
        # 归一化
        output[:, :, i:i_end, :] = output_block / sum_exp
    
    return output


# ============================================================================
# KV Cache 管理
# ============================================================================

class KVCache:
    """
    KV Cache管理器
    
    用于管理PagedAttention的KV缓存。
    """
    
    def __init__(
        self,
        num_blocks: int,
        block_size: int,
        num_kv_heads: int,
        head_dim: int,
        dtype: torch.dtype = torch.float16,
        device: Optional[torch.device] = None,
    ):
        self.num_blocks = num_blocks
        self.block_size = block_size
        self.num_kv_heads = num_kv_heads
        self.head_dim = head_dim
        self.dtype = dtype
        self.device = device
        
        # 分配缓存
        self.key_cache = torch.zeros(
            (num_blocks, block_size, num_kv_heads, head_dim),
            dtype=dtype,
            device=device,
        )
        self.value_cache = torch.zeros(
            (num_blocks, block_size, num_kv_heads, head_dim),
            dtype=dtype,
            device=device,
        )
        
        # 空闲block列表
        self.free_blocks = list(range(num_blocks))
    
    def allocate_block(self) -> int:
        """分配一个block"""
        if not self.free_blocks:
            raise RuntimeError("No free blocks available")
        return self.free_blocks.pop(0)
    
    def free_block(self, block_id: int) -> None:
        """释放一个block"""
        self.free_blocks.append(block_id)
    
    def write_kv(
        self,
        block_id: int,
        slot: int,
        key: torch.Tensor,
        value: torch.Tensor,
    ) -> None:
        """写入KV到指定位置"""
        self.key_cache[block_id, slot] = key
        self.value_cache[block_id, slot] = value
    
    def get_num_free_blocks(self) -> int:
        """获取空闲block数量"""
        return len(self.free_blocks)
